### PR TITLE
added new batching rules and removed all traces of `_chunk_grad_outputs_efficient_attention`

### LINKED ISSUE #102457: Batching rule for `aten::_scaled_dot_product_efficient_attention`
### 🚀 The feature, motivation and pitch

Hi, I am trying to take batched gradients of a vector output given by `_scaled_dot_product_efficient_attention` but saw the error 

```
/site-packages/optimum/bettertransformer/models/attention.py:56: UserWarning: There is a performance drop because we have not yet implemented the batching rul
e for aten::_scaled_dot_product_efficient_attention. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at ../aten/src/ATen/functorch/BatchedFallback.cpp:82.)
  sdpa_result = torch.nn.functional.scaled_dot_product_attention
```

when running my code. Implementing this would really increase throughput in our application! I would also be happy to take a stab at implementing it, if there is a document describing what I need to do at a high level.

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @gchanan @zou3519 @kadeng @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki @Chillee @samdow @kshitij12345 @janeyx99

### PR DESCRIPTION
**Pull request by lengstrom**
Fixes #102457

### CODE REVIEW

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 1]
```diff
@@ -280,6 +280,61 @@ static void expect_at_least_rank(
...
+  bool is_causal,
+  c10::optional<double> scale
+) {
+  // const auto query_ = moveBatchDimToFront(query, query_bdim);
+  // const auto key_ = moveBatchDimToFront(key, key_bdim);
+  // const auto value_ = moveBatchDimToFront(value, value_bdim);
+
+  const auto query_ = reshape_dim_into(*query_bdim, 0, query);
+  const auto key_ = reshape_dim_into(*key_bdim, 0, key);
+  const auto value_ = reshape_dim_into(*value_bdim, 0, value);
```
Comment (zou3519) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:297]: it's possible that the bdims don't exist. This can happen in a situation where we are running vmap over some inputs but not others and we would need to handle this case

File: aten/src/ATen/LegacyBatchingRegistrations.cpp [hunk 2]
```diff
@@ -588,6 +588,17 @@ bool _has_same_storage_numel_batching_rule(const Tensor& self, const Tensor& oth
   return true;
 }
 
+bool _chunk_grad_outputs_efficient_attention_batch_rule(
```
Comment (zou3519) [inline aten/src/ATen/LegacyBatchingRegistrations.cpp:591]: As you've noticed, this file doesn't actually get used by our vmap implementation anymore. It's a relic from the past that we have yet to delete.

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 3]
```diff
@@ -1065,24 +1065,6 @@ Tensor triton_scaled_dot_attention(const Tensor& q, const Tensor& k, const Tenso
 
 REGISTER_CUDA_DISPATCH(_fused_sdp_choice_stub, &_fused_sdp_choice_cuda);
 
-// !!This function is deprecated. See FunctionsManual.cpp for the implementation!!
```
Comment (zou3519) [inline aten/src/ATen/native/transformers/cuda/attention.cu:1068]: It might be better to do the deletion of this function and the addition of the batch rule in separate PRs so that it's easier to review. But totally up to you.

**Review by zou3519 — COMMENTED** [review]
This is definitely going in the right direction! I left some comments.

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp (hunk 1 above)
Comment (lengstrom) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:297]: ahh that makes sense! Can I rely on either all the bdims being existent (or non-existent) together?

Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:297]: No, there might be a mix. You can use `ensure_has_bdim` to take a Tensor and expand out the bdim though, but this sometimes comes at the cost of performance

Comment (lengstrom) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:297]: ok! i will look at other functions to see what the best practice is and try to move in a fix

File: aten/src/ATen/LegacyBatchingRegistrations.cpp (hunk 2 above)
Comment (lengstrom) [reply aten/src/ATen/LegacyBatchingRegistrations.cpp:591]: oops i meant to delete this!

**Comment by lengstrom** [comment]
had to open a new pr here, sorry! https://github.com/pytorch/pytorch/pull/103490
